central/sensor/service/pipeline/complianceoperatorscans/pipeline.go

```diff
@@ -57,30 +67,93 @@ func (s *pipelineImpl) Reconcile(ctx context.Context, clusterID string, storeMap
 	if err := pgutils.RetryIfPostgres(walkFn); err != nil {
 		return err
 	}
+
+	// For now if nextgen compliance is enabled, we have to reconcile both versions of compliance.
+	if features.ComplianceEnhancements.Enabled() {
+		scanObjects, err := s.v2Datastore.GetScanObjectByCluster(ctx, clusterID)
+		if err != nil {
+			return err
+		}
+
+		for _, scanObject := range scanObjects {
+			// The UID is used for reconciliation
+			existingIDs.Add(scanObject.GetId())
+		}
+	}
+
 	store := storeMap.Get((*central.SensorEvent_ComplianceOperatorScan)(nil))
 	return reconciliation.Perform(store, existingIDs, "complianceoperatorscans", func(id string) error {
+		if features.ComplianceEnhancements.Enabled() {
+			if err := s.v2Datastore.DeleteScanObject(ctx, id); err != nil {
+				return err
+			}
+		}
+
 		return s.datastore.Delete(ctx, id)
 	})
 }
 
 func (s *pipelineImpl) Match(msg *central.MsgFromSensor) bool {
+	if features.ComplianceEnhancements.Enabled() {
+		return msg.GetEvent().GetComplianceOperatorScanV2() != nil || msg.GetEvent().GetComplianceOperatorScan() != nil
+	}
+
 	return msg.GetEvent().GetComplianceOperatorScan() != nil
 }
 
 // Run runs the pipeline template on the input and returns the output.
-func (s *pipelineImpl) Run(_ context.Context, clusterID string, msg *central.MsgFromSensor, _ common.MessageInjector) error {
+func (s *pipelineImpl) Run(ctx context.Context, clusterID string, msg *central.MsgFromSensor, _ common.MessageInjector) error {
 	defer countMetrics.IncrementResourceProcessedCounter(pipeline.ActionToOperation(msg.GetEvent().GetAction()), metrics.ComplianceOperatorScan)
 
 	event := msg.GetEvent()
-	scan := event.GetComplianceOperatorScan()
-	scan.ClusterId = clusterID
+
+	// If a sensor sends in a v1 compliance message we will still process it the v1 way in the event
+	// a sensor is not updated or does not have the flag on.
+	switch event.Resource.(type) {
+	case *central.SensorEvent_ComplianceOperatorScan:
+		return s.processComplianceScan(ctx, event, clusterID)
+	case *central.SensorEvent_ComplianceOperatorScanV2:
+		if !features.ComplianceEnhancements.Enabled() {
+			return errors.New("Next gen compliance is disabled.  Message unexpected.")
```

Comment:
Nit: we have a sort of undocumented format to start the message with "UNEXPECTED: "